must-call-checker/src/main/java/org/checkerframework/checker/mustcall/MustCallAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,207 @@
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.util.Elements;
+import org.checkerframework.checker.mustcall.qual.MustCall;
+import org.checkerframework.checker.mustcall.qual.MustCallAny;
+import org.checkerframework.checker.mustcall.qual.PolyMustCall;
+import org.checkerframework.checker.objectconstruction.qual.NotOwning;
+import org.checkerframework.common.basetype.BaseAnnotatedTypeFactory;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
+import org.checkerframework.framework.type.ElementQualifierHierarchy;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
+import org.checkerframework.javacutil.BugInCF;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * The annotated type factory for the must call checker. Primarily responsible for the subtyping
+ * rules between @MustCall annotations.
+ */
+public class MustCallAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+  /** The top annotation. */
+  final AnnotationMirror TOP;
+
+  /** The bottom annotation, which is the default in unannotated code. */
+  final AnnotationMirror BOTTOM;
+
+  /** The polymorphic qualifier */
+  final AnnotationMirror POLY;
+
+  /**
+   * Default constructor matching super. Should be called automatically.
+   *
+   * @param checker the checker associated with this type factory
+   */
+  public MustCallAnnotatedTypeFactory(final BaseTypeChecker checker) {
+    super(checker);
+    TOP = AnnotationBuilder.fromClass(elements, MustCallAny.class);
+    BOTTOM = createMustCall();
+    POLY = AnnotationBuilder.fromClass(elements, PolyMustCall.class);
+    this.postInit();
+  }
+
+  /** Treat non-owning method parameters as @MustCallAny to avoid false positives. */
+  @Override
+  public void methodFromUsePreSubstitution(ExpressionTree tree, AnnotatedExecutableType type) {
+    ExecutableElement declaration;
+    if (tree instanceof MethodInvocationTree) {
+      declaration = TreeUtils.elementFromUse((MethodInvocationTree) tree);
+    } else if (tree instanceof MemberReferenceTree) {
+      declaration = (ExecutableElement) TreeUtils.elementFromTree(tree);
+    } else {
+      throw new BugInCF("unexpected type of method tree: " + tree.getKind());
+    }
+    for (int i = 0; i < type.getParameterTypes().size(); i++) {
+      Element paramDecl = declaration.getParameters().get(i);
+      if (getDeclAnnotation(paramDecl, NotOwning.class) != null) {
+        AnnotatedTypeMirror paramType = type.getParameterTypes().get(i);
+        paramType.replaceAnnotation(TOP);
+        // Descend into a varargs array
```

Comment:
Why does this not descend into parameterized types?

Reply:
The simple answer is that we haven't encountered a case where we needed it yet. Now that I think about it, we probably will eventually.